### PR TITLE
feat(js): add Tally class

### PR DESCRIPTION
**Pull request by beagleknight**
This started as a simple PR to add the tally process to the JS library, but I decided to add the feature while adding the feature.

First, the CHANGELOG:

## Changed

- The `Trustee` class now uses the `Election` class to interact with the log entries
- The `Trustee` class handles everything now when performing both the key ceremony and the tally process.

## Added

- The `Trustee` class now have a `teardown` method that is called automatically to clean a few things. It can be called early to avoid memory leaks if needed.
- The `EventManager` class now handles the `events` stream and exports some useful constants.

## Removed 

- The `KeyCeremony` class has been removed.

Now the extended explanation: when I was working on the tally I realized that there was a lot of duplicated code with the key ceremony, so I decided to simplify it a bit. Thanks to the upcoming sandbox I could test everything, and it works like a charm 💖

### CODE REVIEW

File: decidim-bulletin_board-js/src/client/client.js [hunk 1]
```diff
@@ -43,12 +43,22 @@ export class Client {
   /**
    * Process a key ceremony step sending a signed message.
    *
-   * @param {Object} params - An object that include the following options.
-   *  - {String} signedData - The signed data to be processed.
+   * @param {Object} params - An object that include the data to perform the operation.
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/client/client.js:46]:
```suggestion
   * @param {Object} params - An object that includes the data to perform the operation.
```

File: decidim-bulletin_board-js/src/client/client.js [hunk 2]
```diff
@@ -43,12 +43,22 @@ export class Client {
...
+   * @throws Will throw an error if the request is rejected or the data contains an error.
+   */
+  processKeyCeremonyStep(params) {
+    return this.apiClient.processKeyCeremonyStep(params);
+  }
+
+  /**
+   * Process a tally step sending a signed message.
+   *
+   * @param {Object} params - An object that include the data to perform the operation.
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/client/client.js:57]:
```suggestion
   * @param {Object} params - An object that includes the data to perform the operation.
```

File: decidim-bulletin_board-js/src/client/graphql-client.js [hunk 3]
```diff
@@ -98,4 +99,29 @@ export class GraphQLClient {
 
     return result.data.processKeyCeremonyStep.pendingMessage;
   }
+
+  /**
+   * Process a tally step sending a signed message.
+   *
+   * @param {Object} params - An object that include the following options.
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/client/graphql-client.js:106]:
```suggestion
   * @param {Object} params - An object that includes the following options.
```

File: decidim-bulletin_board-js/src/trustee/event_manager.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+   * Initializes an empty stream of events.
+   *
+   * @constructor
+   */
+  constructor() {
+    this.events = new Subject();
+  }
+
+  /**
+   * Receives a callback function and create a new stream that will call
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/trustee/event_manager.js:21]:
```suggestion
   * Receives a callback function and creates a new stream that will call
```

File: decidim-bulletin_board-js/src/trustee/event_manager.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+   *
+   * @param {fn} Function - A callback function.
+   * @returns {Subscription}
+   */
+  subscribe(fn) {
+    return this.events.pipe(tap(fn)).subscribe();
+  }
+
+  /**
+   * Emits a new message received event through the stream.
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/trustee/event_manager.js:32]:
```suggestion
   * Emits a new "message received" event through the stream.
```

File: decidim-bulletin_board-js/src/trustee/event_manager.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+   */
+  broadcastMessageReceived(message) {
+    this.events.next({
+      type: MESSAGE_RECEIVED,
+      message,
+    });
+  }
+
+  /**
+   * Emits a new message processed event through the stream.
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/trustee/event_manager.js:44]:
```suggestion
   * Emits a new "message processed" event through the stream.
```

File: decidim-bulletin_board-js/src/trustee/trustee.js [hunk 7]
```diff
@@ -1,50 +1,132 @@
...
 export class Trustee {
   /**
    * Initializes the class with the given params.
    *
    * @constructor
    * @param {Object} params - An object that contains the initialization params.
-   *  - {String} id - The trustee identifier.
+   *  - {String} uniqueId - The trustee identifier.
+   *  - {Client} bulletinBoardClient - An instance of the Bulletin Board Client
    *  - {Object} identificationKeys - A object that contains both the public and private key for
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/trustee/trustee.js:20]:
```suggestion
   *  - {Object} identificationKeys - An object that contains both the public and private key for
```

**Review by agustibr — APPROVED** [review]

File: decidim-bulletin_board-js/src/trustee/trustee.js [hunk 8]
```diff
@@ -1,50 +1,132 @@
...
+    await this.processKeyCeremonyStep(message);
+    this.hasSetupKeyCeremony = true;
+    return this.hasSetupKeyCeremony;
+  }
+
+  /**
+   * Starts or continues with the key ceremony process.
+   *
+   * @returns {Promise<Object|undefined>}
+   * @throws An exception is raised if the trustee needs to be restored or
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/trustee/trustee.js:86]: First exception (`!this.hasSetupKeyCeremony`) is missing in here. 😊

File: decidim-bulletin_board-js/src/trustee/trustee.js [hunk 9]
```diff
@@ -80,12 +153,119 @@ export class Trustee {
...
-    return (
+    this.hasSetupKeyCeremony =
       lastMessageFromTrustee &&
-      this.wrapper.restore(wrapperState, lastMessageFromTrustee.messageId)
+      this.wrapper.restore(wrapperState, lastMessageFromTrustee.messageId);
+    return this.hasSetupKeyCeremony;
+  }
+
+  /**
+   * Creates a interval that will check periodically if there are new log entries
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/trustee/trustee.js:166]:
```suggestion
   * Creates an interval that will check periodically if there are new log entries
```

File: decidim-bulletin_board-js/src/trustee/trustee.js [hunk 10]
```diff
@@ -80,12 +153,119 @@ export class Trustee {
...
+          const result = await this.processNextLogEntry();
+          clearInterval(intervalId);
+          resolve(result);
+        }
+      }, this.options.waitUntilNextCheck);
+    });
+  }
+
+  /**
+   * Process the next log entry and outputs the result. It broadcasts an event
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/trustee/trustee.js:188]:
```suggestion
   * Processes the next log entry and outputs the result. It broadcasts an event
```

File: decidim-bulletin_board-js/src/trustee/trustee.js [hunk 11]
```diff
@@ -80,12 +153,119 @@ export class Trustee {
...
+   * @private
+   * @returns {Boolean}
+   */
+  isMessageAlreadyLogged({ message_id }) {
+    const { logEntries } = this.election;
+    return logEntries.find((logEntry) => logEntry.messageId === message_id);
+  }
+
+  /**
+   * Sign a message using the trustee identification keys.
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/trustee/trustee.js:259]:
```suggestion
   * Signs a message using the trustee identification keys.
```

**Review by leio10 — APPROVED** [review]
Good work! 👍

File: decidim-bulletin_board-js/src/trustee/trustee.test.js [hunk 12]
```diff
@@ -29,35 +47,296 @@ describe("Trustee", () => {
   });
 
   it("initialise the trustee with the given params", () => {
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/trustee/trustee.test.js:49]:
```suggestion
  it("initialises the trustee with the given params", () => {
```

File: decidim-bulletin_board-js/src/trustee/trustee.test.js [hunk 13]
```diff
@@ -29,35 +47,296 @@ describe("Trustee", () => {
...
+    );
     expect(trustee.identificationKeys).toEqual(
       defaultParams.identificationKeys
     );
     expect(trustee.election).toEqual(defaultParams.election);
+    expect(trustee.nextLogEntryIndexToProcess).toEqual(0);
+    expect(trustee.lastMessageProcessedWithResult).toBeNull();
   });
 
   it("initialise the trustee wrapper with the given params", () => {
```
Comment (slickepinne) [inline decidim-bulletin_board-js/src/trustee/trustee.test.js:62]:
```suggestion
  it("initialises the trustee wrapper with the given params", () => {
```

**Review by slickepinne — APPROVED** [review]
Nice one. Just a couple of minor english things. 🌻

**Pull request merged**